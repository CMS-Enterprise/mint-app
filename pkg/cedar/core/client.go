package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
