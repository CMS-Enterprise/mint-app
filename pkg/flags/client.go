package flags

import (
	"context"
	"crypto/sha256"
	"fmt"
	"time"

	"github.com/cmsgov/mint-app/pkg/appcontext"

	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfiledata"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfilewatch"

	"github.com/cmsgov/mint-app/pkg/appconfig"
)

// DowngradeAssessmentTeamKey is the Flag Key in LaunchDarkly for downgrading assessment team users
const DowngradeAssessmentTeamKey = "downgradeAssessmentTeam"

// Config has all the parts for creating a new LD Client
type Config struct {
	Source         appconfig.FlagSourceOption
	Key            string
	Timeout        time.Duration
	FlagValuesFile string
}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	if config.Source == appconfig.FlagSourceLaunchDarkly {
		return ld.MakeClient(config.Key, config.Timeout)
	}

	defaultTimeout := 5 * time.Second

	if config.Source == appconfig.FlagSourceFile {

		ldConfig := ld.Config{
			DataSource: ldfiledata.DataSource().FilePaths(config.FlagValuesFile).Reloader(ldfilewatch.WatchFiles),
		}
		return ld.MakeCustomClient("fake_key", ldConfig, defaultTimeout)
	}

	// we default to an OFFLINE client for non-deployed environments
	return ld.MakeCustomClient("fake_offline_key", ld.Config{Offline: true}, 5*time.Second)
}

// Principal builds the LaunchDarkly user object for the
// currently authenticated principal.
func Principal(ctx context.Context) lduser.User {
	p := appcontext.Principal(ctx)
	key := UserKeyForID(p.ID())

	// this is a bit of a loose inference, assuming a user w/o Job Codes
	// is an Anonymous user. Over time, may want to consider adding
	// a `func Anonymous() bool` accessor to the authorizaion.Principal interface
	// definition instead of doing this inference
	authed := (p.AllowUSER() || p.AllowASSESSMENT())

	return lduser.
		NewUserBuilder(key).
		Anonymous(!authed).
		Build()
}

// UserKeyForID generates a user key from an ID
// we should not be using bare EUA IDs as identifiers to
// LaunchDarkly (per Jimil/ISSO), so we use a cryptographically
// secure one-way hash of the EUA ID as "key" for the LD User object.
func UserKeyForID(id string) string {
	h := sha256.New()
	_, _ = h.Write([]byte(id))
	return fmt.Sprintf("%x", h.Sum(nil))
}
