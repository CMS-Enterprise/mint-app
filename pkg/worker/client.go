package worker

import (
	"fmt"

	faktory "github.com/contribsys/faktory/client"
)

// NewClient returns a new faktory Client.
func NewClient() (*faktory.Client, error) {
	cl, err := faktory.Open()
	if err != nil {
		return nil, fmt.Errorf("error opening faktory client: %w", err)
	}
	return cl, nil
}
