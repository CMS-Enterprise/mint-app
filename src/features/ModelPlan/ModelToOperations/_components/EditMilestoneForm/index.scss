.save-tag {
    display: flex;
    flex-direction: column;
    align-items: end;
    padding-right: 2rem;
    margin-top: -3rem;
    margin-bottom: .9rem;
}