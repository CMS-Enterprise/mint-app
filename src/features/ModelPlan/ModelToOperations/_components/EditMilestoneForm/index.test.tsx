import React from 'react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen, waitFor } from '@testing-library/react';
import {
  allMTOSolutionsMock,
  categoryMock,
  milestoneMock,
  modelID
} from 'tests/mock/mto';

import MessageProvider from 'contexts/MessageContext';

import EditMilestoneForm from '.';

describe('EditMilestoneForm', () => {
  it('matches snapshot', async () => {
    const router = createMemoryRouter(
      [
        {
          path: '/models/:modelID/collaboration-area/model-to-operations/matrix',
          element: (
            <EditMilestoneForm
              closeModal={() => {}}
              setIsDirty={() => {}}
              setCloseDestination={vi.fn()}
              setFooter={() => {}}
              submitted={{ current: false }}
            />
          )
        }
      ],
      {
        initialEntries: [
          `/models/${modelID}/collaboration-area/model-to-operations/matrix?view=milestones&edit-milestone=123`
        ]
      }
    );

    const { asFragment } = render(
      <MessageProvider>
        <MockedProvider
          mocks={[
            ...milestoneMock('123'),
            ...categoryMock,
            ...allMTOSolutionsMock
          ]}
          addTypename={false}
        >
          <RouterProvider router={router} />
        </MockedProvider>
      </MessageProvider>
    );

    await waitFor(() => {
      expect(screen.getByText('Milestone 1')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
