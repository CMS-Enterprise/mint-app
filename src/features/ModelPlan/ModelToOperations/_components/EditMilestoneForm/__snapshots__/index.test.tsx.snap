// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditMilestoneForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="margin-top-8"
  >
    <div
      class="save-tag"
    >
      <div
        class="bg-warning-lighter padding-y-05 padding-x-1"
      >
        <svg
          aria-label="warning"
          class="usa-icon margin-right-1 top-2px text-warning"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
        </svg>
        <p
          class="margin-0 display-inline margin-right-1"
        >
          1 unsaved change
        </p>
        -
        <button
          class="usa-button usa-button--unstyled margin-x-1"
          data-testid="button"
          type="button"
        >
          Save
        </button>
      </div>
    </div>
    <div
      class="padding-8 maxw-tablet"
    >
      <form
        class="usa-form maxw-none"
        data-testid="form"
        id="edit-milestone-form"
      >
        <h2
          class="margin-y-2 margin-bottom-4 padding-bottom-4 line-height-large border-bottom-1px border-base-lighter"
        >
          Milestone 1
        </h2>
        <fieldset
          class="usa-fieldset margin-bottom-8"
          data-testid="fieldset"
        >
          <p
            class="margin-top-0 margin-bottom-3 text-base"
          >
            Fields marked with an asterisk ( 
            <span
              class="text-secondary-dark"
            >
              *
            </span>
             ) are required.
          </p>
          <div
            class="usa-form-group margin-top-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <div
              class="mint-checkbox usa-checkbox"
            >
              <input
                class="usa-checkbox__input"
                id="isDraft"
                name="isDraft"
                type="checkbox"
                value="isDraft"
              />
              <label
                class="usa-checkbox__label"
                for="isDraft"
              >
                Check if this milestone is a draft milestone
              </label>
              <p
                class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
              >
                The "draft" indicator will signal to others that this milestone is more of a work in progress than the rest of the model-to-operations matrix (MTO).
              </p>
            </div>
          </div>
          <div
            class="usa-form-group margin-top-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <label
              class="usa-label maxw-none text-bold"
              data-testid="label"
              for="categories.category.id"
            >
              Category 
              <abbr
                class="usa-hint usa-hint--required"
                title="required"
              >
                *
              </abbr>
            </label>
            <div
              class="mint-help-text margin-top-05"
            >
              If you don’t see what you’re looking for, you may add a custom category from the main page of your MTO.
            </div>
            <select
              class="usa-select"
              data-testid="Select"
              id="categories.category.id"
              name="categories.category.id"
            >
              <option
                value="default"
              >
                - Select - 
              </option>
              <option
                value="123"
              >
                Category 1
              </option>
              <option
                value="456"
              >
                Category 2
              </option>
            </select>
          </div>
          <div
            class="usa-form-group margin-top-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <label
              class="usa-label maxw-none text-bold"
              data-testid="label"
              for="categories.sub-category.id"
            >
              Sub-category 
              <abbr
                class="usa-hint usa-hint--required"
                title="required"
              >
                *
              </abbr>
            </label>
            <div
              class="mint-help-text margin-top-05"
            >
              If you don’t see what you’re looking for, you may add a custom category from the main page of your MTO.
            </div>
            <select
              class="usa-select"
              data-testid="Select"
              id="categories.sub-category.id"
              name="categories.subCategory.id"
            >
              <option
                value="default"
              >
                - Select - 
              </option>
            </select>
          </div>
          <div
            class="usa-form-group margin-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <label
              class="usa-label maxw-none text-bold"
              data-testid="label"
              for="responsible-component"
            >
              Responsible component 
              <abbr
                class="usa-hint usa-hint--required"
                title="required"
              >
                *
              </abbr>
            </label>
            <div
              class="mint-help-text margin-top-1"
            >
              Choose the CMS component or CMMI group responsible for this work. Select all that apply.
            </div>
            <div>
              <div
                class="easi-multiselect usa-combo-box margin-top-1 css-b62m3t-container"
                id="responsible-component"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-2-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                  role="log"
                />
                <div
                  class=" css-1vl89cy-control"
                >
                  <div
                    class=" css-1fdsijx-ValueContainer"
                  >
                    <div
                      class=" css-1jqq78o-placeholder"
                      id="react-select-2-placeholder"
                    >
                      0 selected
                    </div>
                    <div
                      class=" css-qbdosj-Input"
                      data-value=""
                    >
                      <input
                        aria-activedescendant=""
                        aria-autocomplete="list"
                        aria-describedby="react-select-2-placeholder"
                        aria-expanded="false"
                        aria-haspopup="true"
                        aria-label="Responsible component"
                        aria-labelledby="responsible-component"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class=""
                        id="responsible-component"
                        role="combobox"
                        spellcheck="false"
                        style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class=" css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class=" css-16b81-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-gjng1h-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div>
                  <input
                    name="responsibleComponent"
                    type="hidden"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="usa-form-group margin-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <label
              class="usa-label"
              data-testid="label"
              for="facilitated-by"
            >
              Facilitated by
            </label>
            <div
              class="mint-help-text margin-top-1"
            >
              Choose the role or group responsible for this work.
            </div>
            <div>
              <div
                class="easi-multiselect usa-combo-box margin-top-1 css-b62m3t-container"
                id="multi-source-data-to-collect"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-3-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                  role="log"
                />
                <div
                  class=" css-1vl89cy-control"
                >
                  <div
                    class=" css-1fdsijx-ValueContainer"
                  >
                    <div
                      class=" css-1jqq78o-placeholder"
                      id="react-select-3-placeholder"
                    >
                      0 selected
                    </div>
                    <div
                      class=" css-qbdosj-Input"
                      data-value=""
                    >
                      <input
                        aria-activedescendant=""
                        aria-autocomplete="list"
                        aria-describedby="react-select-3-placeholder"
                        aria-expanded="false"
                        aria-haspopup="true"
                        aria-label="Facilitated by"
                        aria-labelledby="facilitated-by"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class=""
                        id="facilitated-by"
                        role="combobox"
                        spellcheck="false"
                        style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class=" css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class=" css-16b81-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-gjng1h-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div>
                  <input
                    name="facilitatedBy"
                    type="hidden"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="usa-form-group margin-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <label
              class="usa-label"
              data-testid="label"
              for="need-by"
            >
              Need by
            </label>
            <div
              class="mint-help-text margin-top-1"
            >
              Choose the date when onboarding and implementation work for all solutions in this milestone should be complete. Format: mm/dd/yyyy
            </div>
            <div
              class="position-relative"
            >
              <div
                class="usa-date-picker usa-date-picker--initialized"
                data-testid="date-picker"
              >
                <input
                  aria-hidden="true"
                  aria-labelledby="need-by"
                  class="usa-input usa-sr-only usa-date-picker__internal-input"
                  data-testid="date-picker-internal-input"
                  name="needBy"
                  readonly=""
                  tabindex="-1"
                  type="text"
                  value="2021-08-01T00:00:00.000Z"
                />
                <div
                  class="usa-date-picker__wrapper"
                  tabindex="-1"
                >
                  <input
                    aria-labelledby="need-by"
                    class="usa-input usa-date-picker__external-input"
                    data-testid="date-picker-external-input"
                    id="milestone-need-by"
                    type="text"
                    value="08/01/2021"
                  />
                  <button
                    aria-haspopup="true"
                    aria-label="Toggle calendar"
                    class="usa-date-picker__button"
                    data-testid="date-picker-button"
                    type="button"
                  />
                  <div
                    class="usa-date-picker__calendar"
                    data-testid="date-picker-calendar"
                    hidden=""
                    role="application"
                    style="top: 0px;"
                  />
                  <div
                    aria-live="polite"
                    class="usa-sr-only usa-date-picker__status"
                    data-testid="date-picker-status"
                    role="status"
                  />
                </div>
              </div>
              <div
                class="mint-datepicker-warning"
              >
                <span
                  class="usa-tooltip"
                  data-testid="tooltipWrapper"
                >
                  <div
                    aria-describedby="tooltip-«r0»"
                    class="usa-tooltip__trigger"
                    data-testid="triggerElement"
                    tabindex="0"
                    title=""
                  >
                    <svg
                      aria-label="warning"
                      class="usa-icon usa-icon--size-3 text-warning-dark"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
                      />
                    </svg>
                  </div>
                  <span
                    aria-hidden="true"
                    class="usa-tooltip__body"
                    data-testid="tooltipBody"
                    id="tooltip-«r0»"
                    role="tooltip"
                    title="You’ve entered one or more dates that are in the past. Please double-check your dates to make sure they’re accurate."
                  >
                    You’ve entered one or more dates that are in the past. Please double-check your dates to make sure they’re accurate.
                  </span>
                </span>
              </div>
            </div>
            <div
              class="usa-alert usa-alert--warning usa-alert--slim flex margin-top-2"
              data-testid="alert"
            >
              <div
                class="usa-alert__body"
              >
                <p
                  class="usa-alert__text"
                >
                  <span>
                    You’ve entered one or more dates that are in the past. Please double-check your dates to make sure they’re accurate.
                  </span>
                </p>
              </div>
            </div>
          </div>
          <div
            class="usa-form-group margin-top-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <label
              class="usa-label maxw-none text-bold"
              data-testid="label"
              for="status"
            >
              Status 
              <abbr
                class="usa-hint usa-hint--required"
                title="required"
              >
                *
              </abbr>
            </label>
            <select
              class="usa-select"
              data-testid="Select"
              id="status"
              name="status"
            >
              <option
                value="NOT_STARTED"
              >
                Not started
              </option>
              <option
                value="IN_PROGRESS"
              >
                In progress
              </option>
              <option
                value="COMPLETED"
              >
                Completed
              </option>
            </select>
          </div>
          <div
            class="desktop:grid-col-12 margin-bottom-4"
            data-testid="grid"
          >
            <button
              class="usa-button usa-button--unstyled display-flex flex-align-center text-ls-1 deep-underline margin-bottom-1 margin-top-3"
              data-testid="toggle-need-answer"
              type="button"
            >
              <svg
                aria-label="collapse"
                class="usa-icon margin-right-05"
                focusable="false"
                height="1em"
                role="img"
                style="transform: rotate(90deg);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
                />
              </svg>
              Milestone implementation statuses
            </button>
          </div>
          <div
            class="usa-form-group margin-top-0 margin-bottom-3"
            data-testid="formGroup"
          >
            <label
              class="usa-label maxw-none text-bold"
              data-testid="label"
              for="risk-indicator"
            >
              Risk indicator 
              <abbr
                class="usa-hint usa-hint--required"
                title="required"
              >
                *
              </abbr>
            </label>
            <div
              class="mint-help-text margin-top-1"
            >
              Select the risk level for this milestone. This will help you and your team identify potential risks and plan accordingly.
            </div>
            <div
              class="display-flex"
            >
              <div
                class="usa-radio margin-right-1"
                data-testid="radio"
              >
                <input
                  checked=""
                  class="usa-radio__input"
                  id="risk-indicator-ON_TRACK"
                  name="riskIndicator"
                  type="radio"
                  value="ON_TRACK"
                />
                <label
                  class="usa-radio__label"
                  for="risk-indicator-ON_TRACK"
                >
                  No risk (on track)
                </label>
              </div>
            </div>
            <div
              class="display-flex"
            >
              <div
                class="usa-radio margin-right-1"
                data-testid="radio"
              >
                <input
                  class="usa-radio__input"
                  id="risk-indicator-OFF_TRACK"
                  name="riskIndicator"
                  type="radio"
                  value="OFF_TRACK"
                />
                <label
                  class="usa-radio__label"
                  for="risk-indicator-OFF_TRACK"
                >
                  Some risk (off track)
                </label>
              </div>
              <svg
                aria-label="warning"
                class="usa-icon usa-icon--size-3 text-warning-dark"
                focusable="false"
                height="1em"
                role="img"
                style="top: 10px;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
                />
              </svg>
            </div>
            <div
              class="display-flex"
            >
              <div
                class="usa-radio margin-right-1"
                data-testid="radio"
              >
                <input
                  class="usa-radio__input"
                  id="risk-indicator-AT_RISK"
                  name="riskIndicator"
                  type="radio"
                  value="AT_RISK"
                />
                <label
                  class="usa-radio__label"
                  for="risk-indicator-AT_RISK"
                >
                  Significantly at risk
                </label>
              </div>
              <svg
                aria-label="error"
                class="usa-icon usa-icon--size-3 text-error-dark"
                focusable="false"
                height="1em"
                role="img"
                style="top: 10px;"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </svg>
            </div>
          </div>
          <div
            class="border-top-1px border-base-lighter padding-y-4"
          >
            <h3
              class="margin-0 margin-bottom-1"
            >
              Selected solutions
            </h3>
            <p
              class="margin-0 margin-bottom-1"
            >
              1 solutions associated with this milestone
            </p>
            <button
              class="usa-button usa-button--unstyled margin-0 display-flex"
              data-testid="button"
              type="button"
            >
              Edit solutions
              <svg
                aria-label="forward"
                class="usa-icon top-2px"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                />
              </svg>
            </button>
            <table
              class="usa-table usa-table--borderless width-full margin-top-0"
              data-testid="table"
            >
              <thead>
                <tr
                  role="row"
                >
                  <th
                    class="padding-left-0 padding-bottom-0"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="width: auto;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Solution
                      <svg
                        aria-label="unfold"
                        class="usa-icon margin-left-05"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    class="padding-left-0 padding-bottom-0"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="width: 150px;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                      type="button"
                    >
                      Status
                      <svg
                        aria-label="unfold"
                        class="usa-icon margin-left-05"
                        data-testid="caret--sort"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                        />
                      </svg>
                    </button>
                  </th>
                  <th
                    class="padding-left-0 padding-bottom-0"
                    colspan="1"
                    role="columnheader"
                    scope="col"
                    style="width: auto;"
                  >
                    <button
                      class="usa-button usa-button--unstyled position-relative"
                      type="button"
                    >
                      <svg
                        aria-label="warning"
                        class="usa-icon usa-icon--size-3 left-05 text-base-lighter"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
                        />
                      </svg>
                    </button>
                  </th>
                </tr>
              </thead>
              <tbody
                role="rowgroup"
              >
                <tr
                  role="row"
                >
                  <td
                    class="padding-left-0"
                    role="cell"
                  >
                    Solution 1
                  </td>
                  <td
                    class="padding-left-0"
                    role="cell"
                  >
                    <div
                      class="model-plan-task-list__task-tag line-height-body-1 text-bold bg-success-dark text-white width-fit-content"
                      data-testid="tasklist-tag"
                    >
                      <span>
                        Completed
                      </span>
                    </div>
                  </td>
                  <td
                    class="padding-left-0"
                    role="cell"
                  >
                    <div>
                      <span
                        class="usa-tooltip"
                        data-testid="tooltipWrapper"
                      >
                        <div
                          aria-describedby="tooltip-«r1»"
                          class="usa-tooltip__trigger"
                          data-testid="triggerElement"
                          tabindex="0"
                          title=""
                        >
                          <svg
                            class="usa-icon usa-icon--size-3 text-error-dark top-05"
                            focusable="false"
                            height="1em"
                            role="img"
                            viewBox="0 0 24 24"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                            />
                          </svg>
                        </div>
                        <span
                          aria-hidden="true"
                          class="usa-tooltip__body"
                          data-testid="tooltipBody"
                          id="tooltip-«r1»"
                          role="tooltip"
                          title="Significantly at risk"
                        >
                          Significantly at risk
                        </span>
                      </span>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="usa-alert usa-alert--info usa-alert--slim flex margin-top-4"
              data-testid="alert"
            >
              <div
                class="usa-alert__body"
              >
                <p
                  class="usa-alert__text"
                >
                  <span>
                    To view more information about the solutions associated with this milestone, navigate to the 
                    <button
                      class="usa-button usa-button--unstyled usa-button--unstyled margin-0"
                      data-testid="button"
                      type="button"
                    >
                      “Solutions and IT systems” tab
                    </button>
                     in your MTO. From there you may edit details about the solutions, including the status.
                  </span>
                </p>
              </div>
            </div>
          </div>
          <div
            class="border-top-1px border-base-lighter padding-y-4"
          >
            <div>
              <h3
                class="margin-0 margin-bottom-1"
              >
                Milestone notes
              </h3>
              <p
                class="margin-0 margin-bottom-1"
              >
                0 notes added to this milestone
              </p>
              <button
                class="usa-button usa-button--unstyled margin-0 display-flex"
                data-testid="add-note-button"
                type="button"
              >
                Add a note
                <svg
                  aria-label="forward"
                  class="usa-icon top-2px"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
